### ISSUE TITLE
Prepare to remove prebuilt LLVM binaries. [phab:D128]

### ISSUE THREAD

**Issue by kythe** [google/kythe#333]
_Created by **zarko** at 2015-03-10 23:08:01_: 

This CL is the first in a series that will remove
prebuilt LLVM binaries and require developers to
point campfire toward LLVM on their own. It introduces
three new properties to .campfire_settings:

llvm_repo, a path to an LLVM checkout, used to make sure
  that the version is up to date
llvm_lib, a path (relative to third_party/llvm) to use
  to look for built LLVM libraries
llvm_include, a path or array of paths (relative to the
  build root) to add to the include search list.

buildtools/check_modules.sh uses the llvm_repo
property to interrogate LLVM checkouts. It supports
git or svn. In this CL, the assertions about minimum
checkout versions are disabled. This script is run
each time campfire is invoked.

I have also amended some extractor tests to make them
more robust in the face of changing compiler resource
files.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

This works fine, but rubs me the wrong way. I'd be happier if there were a new variable for each of these transforms (or they were iterated), e.g.,

   var barePaths = includePaths.reduce(…);
   var fixedPaths = barePaths.reduce(…);

or `.reduce(…).reduce(…)`.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

Surely: `/bin/bash` ?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

I assume these will need to be updated in pairs, then?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

Optional:

    "${1:?missing repo path}"
    "${2:?missing friendly name}"
    "${3:?missing SHA}"

etc. for one (perhaps the first) of the occurrences of each.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

s/properites/properties/

Where is `s` defined? (Is that `instring` ?)

I think the Properites may have been a Moabite sect, though.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

I think what this means is that `substituteArrayProperties(v)` on a property p with values `[1,2,3]` will yield `[s:p→1, s:p→2, s:p→3]` ?

What are the substituents?
Maybe an example would make this easier to document.

Edit: Having read further, I know now what the substituents look like. But please document them (briefly) anyway.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

Or, if you like

```
var changed = false;
do {
   …
} while (changed);
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-11 03:39:00_:

I feel like all of this would be more compact as, e.g., 

   return instring.replace(/\$\((\w+)\)/g, function() {
       …something interesting with arguments[1]…
   });

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-11 17:35:20_:

Or `/bin/sh`

**Comment by kythe** [comment]
_**zarko** wrote at 2015-03-11 18:10:31_:

  - Address comments.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

Thanks; I meant `/bin/sh`. This was `dash` when I was testing for bashisms.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

Yes--it's easy to do because LLVM uses git-svn, so git SHAs from origin map to svn revisions.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

Done

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

Done.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

Done.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

Done

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

This preserves the existing substitution behavior (substituteProperties was extracted from code elsewhere). Since we're going from strings to arrays here, it's unclear that .replace() will help.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-11 18:12:36_:

Done.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-03-11 23:15:18_:

  - Compatibility.
  - Don't check modules on non-buildy actions.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-03-12 18:28:27_:

  - When merging, treat arrays as values, not as dictionaries.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-03-12 22:57:24_:

ping

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-03-12 23:21:01_:

Could you add a line or two summarizing how these get updated?

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2015-03-12 23:21:01_

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-12 23:21:53_:

Can we rename this to llvm_relative_lib or something similar to connote its relativity?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-12 23:21:53_:

Why isn't this relative as well?

Also, why can't these be derived from just llvm_repo?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-12 23:26:54_:

Isn't this relative to llvm_repo?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-12 23:26:54_:

Ditto. This script is highly cryptic.  Why do we even have to check the versions?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-12 23:54:32_:

It is not. (You might build in a completely different directory.)

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-12 23:54:32_:

If you don't think this is necessary, we can just remove the script. I figured that it would be better for developers to get a warning printed when our LLVM version expectations change.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-12 23:54:32_:

Do you mean how the strings in the file get updated, or how the underlying repo gets updated?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-12 23:54:32_:

The general expectation I have is that these will either be untouched or they will be set to (possibly unrelated) absolute paths. In the CL with the prebuilt libs removed, this looks like:

```
      "llvm_repo": "third_party/llvm/llvm",
      "llvm_lib": "llvm/build/Release+Asserts/lib",
      "llvm_include": [
        "third_party/llvm/llvm/include",
        "third_party/llvm/llvm/tools/clang/include",
        "third_party/llvm/llvm/build/include",
        "third_party/llvm/llvm/build/tools/clang/include"
      ],
```

I'm not sure that it's necessary that llvm_lib has the same implicit root as llvm_include--this implementation turned out to require the fewest changes to campfire. I'd appreciate any suggestions on how to simplify it further (especially since you are more familiar with the build code than I am).

I am particularly hesitant to make llvm_* implicitly relative to llvm_repo, since that seems like it would require special-casing the property names or making absolute paths weird to deal with.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-13 14:56:56_:

Both would be helpful, but I think he meant these particular strings.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-03-13 18:22:35_:

  - Address comments.
  - Compatibility.
  - Don't check modules on non-buildy actions.
  - When merging, treat arrays as values, not as dictionaries.
  - Make property names clearer.

**Comment by kythe** [comment]
_**zarko** closed via commit 783f30439f3a7b3b90486e80aebbffb91ad768df at 2015-03-13 18:23:00_
